byterun/intern.c

```diff
@@ -718,15 +720,25 @@ value caml_input_val(struct channel *chan)
   }
   /* Initialize global state */
   intern_init(block, block);
-  intern_alloc(h.whsize, h.num_objects);
+  intern_alloc(h.whsize, h.num_objects, outside_heap);
   /* Fill it in */
   intern_rec(&res);
-  intern_add_to_heap(h.whsize);
+  if (!outside_heap) {
+    intern_add_to_heap(h.whsize);
+  } else {
+    intern_extra_block = NULL;
+    intern_block = 0;
+  }
```

Comment:
The calling protocol in memory.c is that `caml_alloc_for_heap` is supposed to be followed by either `caml_add_to_heap` or `caml_free_for_heap`. I find it troubling that there is no `free_for_heap` call in the `else` branch here -- it looks like a violation of the protocol, although it may not be a bug if this is somehow called later. Also, to avoid mixing abstraction levels, this should in fact call a `intern_free_for_heap` function that itself calls `caml_free_for_heap` and does the necessary `intern_*` handling.